### DISCUSSION TITLE
What should be the minimum key size in openssl 3.0.x

### DISCUSSION THREAD

**Discussion by manoj4617** [openssl/openssl#22447] [Q&A] [answered]
What should be the minimum key size to be passed to this function `EVP_PKEY_CTX_set_rsa_keygen_bits` , I used to pass 256 as the key size while using openssl 1.0

**Comment by levitte** [comment]
The minimum number of RSA modulus bits (which is what you specify with that function) has been raised to 512.  This happened in the development of OpenSSL 1.1.0, in commit c56a50b229932d2cef651d931b71a8cbffb029da

**Reply by manoj4617** [reply]
Thanks!!

**Comment by mattcaswell** [comment]
> I used to pass 256

256 is *way* too small for RSA. Although the minimum is 512 you should consider setting this to at least 2048.

**Reply by manoj4617** [reply]
Thanks!!

**Comment by tom-cosgrove-arm** [comment, marked as answer]
The "security strength" equivalent you get with RSA is much lower than the actual number of bits in the RSA modulus - 1024-bit RSA has roughly an 80-bit security strength, and 2,048-bit RSA has about 112 bits' security strength. For comparison, 128-bit AES has 128-bit security strength, and 256-bit AES has 256-bit security strength. See table 2 on page 54 of [NIST's Recommendation for Key Management:](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf)

Alternatively, there's a nice summary at https://www.keylength.com/en/4/

In short: these days one should use at least 2,048 bits if using RSA (but most cryptographers prefer elliptic curve algorithms now, if we ignore the issue of post-quantum)

**Reply by manoj4617** [reply]
Thanks
